### ISSUE TITLE
List of companies using KafkaJS

### ISSUE THREAD

**Issue by Nevon** [tulios/kafkajs#289]
**Hey KafkaJS users!**

We sometimes get questions from people on whether or not KafkaJS is used in production. As the maintainers, we hear from people every now and then, but it's not documented anywhere. Knowing how KafkaJS is used and by whom helps us make decisions and also helps users that feel unsure if they can trust this library or not.

✋ **If you or your company is using KafkaJS, please let us know in this thread!** ✋

Once we have a couple of users listed, I will turn this into a PR adding the users section to the website.

## Known users:

* [Klarna](https://klarna.com) - European fintech using KafkaJS across a few dozen services. @tulios & @Nevon

**Comment by kpala** [comment]
[Nightingale Health](https://nightingalehealth.com/) - Finnish health tech company using KafkaJS in a laboratory information management system. @kpala

**Comment by pimpelsang** [comment]
Thinking about it in pipedrive, Tallinn, so far only in tests :)

**Comment by paambaati** [comment]
[Indix](https://indix.com) uses KafkaJS to move massive amounts of data across various systems every day! We also open-sourced the [kafkajs-lz4](https://github.com/indix/kafkajs-lz4) module that we use in-house.

**Comment by vykimo** [comment]
[IBM France Watson](http://www.ibminteractive.fr/#/home)
We use it to handle the pipeline of a cognitive platform. 🙂

**Comment by drojas** [comment]
[Azlo](https://www.azlo.com/) is using KafkaJS to deliver data between from micro-services and from micro-services to data lake

**Comment by julienvincent** [comment]
[https://journeyapps.com/](JourneyApps)

We are now using KafkaJS in three areas:

+ To power communication and data-replication between our many micro-services.
+ To power GraphQL subscriptions for real-time communication between client interfaces
+ To power real-time co-editing in our platform IDE

**Comment by StevenLangbroek** [comment]
We ([Quandoo](https://www.quandoo.com/) are writing our first service with it now.

**Comment by yarncraft** [comment]
As a data engineer at [Charp](http://www.charp.be/), I am using Kafka.js with Confluent to build an IoT streaming analytical layer.

**Comment by StevenLangbroek** [comment]
Oh wow totally forgot to update; yeah we've been running this in production for about 4-5 months now :).

**Comment by ankon** [comment]
Good one, also should mention this here:

We ([NEXT Technologies](https://www.nextapp.co/)) are using KafkaJS to power our platform of about 40 microservices written in TypeScript (event sourcing). We use Kafka hosted in AWS (MSK).

**Comment by WaleedAshraf** [comment]
Hi everyone,

We at relayr are also planning to switch from **Blizzard/node-rdkafka** to **kafka-node** OR **KafkaJS**. We are looking for something which is production-ready.  **Blizzard/node-rdkafka** has many issues with the consumer getting disconnected and weird error codes and no proper way to disconnect and shutdown the producer/consumer and much more.
We have 20+ Nodejs services that produce and consume messages. We also produce and consume in batch. We are looking for something which also provides some kind of back-pressure or at least the ability to implement it. 

If anyone has already done research on pros/cons OR how KafkaJS is a better option, OR if you have faced any issues with **kafka-node** please share. That would be really helpful.

Thanks,

**Comment by kpala** [comment]
> If anyone has already done research on pros/cons OR how KafkaJS is a better option, OR if you have faced any issues with **kafka-node** please share. That would be really helpful.

We had **kafka-node** in use for maybe a year or two since it was the recommended library for Node back then. Lots of connectivity issues and ultimately the only way we could handle these was restarting our services.

Moreover, **kafka-node** provided multiple options for creating producers and consumers and these options just kept piling up over time without a proper insight which of these we were supposed to be using (we started from HighLevelConsumer and ended up to ConsumerGroupStreams). Now and then I had to go read the source code to understand which options were available since this was not properly documented. This also made me realize the codebase was a mess.

When we were finally fed up with the connectivity issues, we started to migrate to **node-rdkafka** with equally disappointing results and since we didn't invest much in that migration, we decided to look for yet another library.

With **kafkajs** our system has been surprisingly stable and we've been happy so far.

However, it's also worth mentioning that during this time we've been improving our infrastructure and Kafka itself has improved a lot as well. Thus, it's not so black-and-white but I wouldn't go back to **kafka-node** nevertheless.

**Comment by WaleedAshraf** [comment]
That's really helpful @kpala. I also found some more info, like [these benchmarks](https://github.com/tulios/kafkajs/issues/398).
Once I'm done with research and comparison, I'll also share my findings here. 👍

**Comment by ThisIsMissEm** [comment]
Just to further on @StevenLangbroek's comment, at Quandoo we've also written a bit of a layer on top of kafkajs to help with managing subscriptions to topics, as we've internal standards for our kafka data.

There are still things that we wish were better documented or easier to understand (e.g., a recipes section, like, how do I calculate consumer lag, how do I know when my consumer is "ready" (kappa), and there's a few issues we had around consumer group names getting reused on different topics which leads to only one topic being subscribed)

But overall, kafkajs has been a pleasure to work with & the support from developers outstanding.

**Comment by WaleedAshraf** [comment]
@ThisIsMissEm for 
>  issues we had around consumer group names getting reused on different topics which leads to only one topic being subscribed

Can you explain a bit more about the issue and how you fixed it?
In our case, we have one consumer-group per service which consumes from different topics. So, we may also run into the same problem if it only consumes from one topic.

**Comment by ThisIsMissEm** [comment]
@WaleedAshraf it was because we had multiple distinct kafkajs consumers, so called subscribe() multiple times with the same groupId, but different topics.

The solution was to use a unique group per topic as a stop-gap, and then write a consumer manager layer, so we could have different parts of code each connect into the kafka instance, but only one actually subscription.

Writing the I have just realised I could probably instantiate my consumer classes separately, and then pass them into a single "kafka subscriber" method or something (It's a slight inversion of control). I'll try writing it tomorrow maybe.

Basically if you have multiple `kafka.consumer`'s then don't share the groupId; also, you'll notice it not working pretty quickly because you'll see it in the logs.

Which reminds me, I should open-source our small function to reformat kafkajs logs for pino.

**Comment by WaleedAshraf** [comment]
Thanks for sharing this. 👍👍 I'll test it out tomorrow. Probably we'll have same issue because we use 1 groupId for multiple topics when we subscribe to consumer.

I'm not sure if there's an issue/pr for this to fix within kafkaJS or why there is this limitation.

**Comment by ThisIsMissEm** [comment]
It's not really a kafkajs issue; it was me misusing the library.

```
const { Kafka } = require('kafkajs')

const kafka = new Kafka({
  clientId: 'my-app',
  brokers: ['kafka1:9092', 'kafka2:9092']
})

const consumer1 = kafka.consumer({ groupId: 'test-group' })

await consumer1.connect()
await consumer1.subscribe({ topic: 'test-topic', fromBeginning: true })

const consumer2 = kafka.consumer({ groupId: 'test-group' })

await consumer2.connect()
await consumer2.subscribe({ topic: 'another-topic', fromBeginning: true })
```

Two kafka.consumers, both sharing the same groupId, but different subscriptions

**Comment by tulios** [comment]
@WaleedAshraf this is not a KafkaJS limitation, consumer groups can subscribe to multiple topics, but all consumers within the group have to subscribe to the same topics. The concept of a consumer group falls short if every consumer in the group has a different subscription.

If you need help with other questions or need help in general, please, create a specific issue or join out slack channel, let's not flood unrelated issues will all sorts of questions.

Thanks.

**Comment by DevNM21** [comment]
[StudyMonk](https://studymonk.in) now uses  KafkaJs to track various events and actions performed by the students in our platform

**Comment by cwnga2** [comment]
We (Oriente) are writing our first service with it now.

**Comment by AustinHunt** [comment]
[HailTrace](https://hailtrace.com)

We love, love KafkaJS <3

**Comment by angelozerr** [comment]
@Nevon IMHO I think you should add those projects in your README (we did that for an another project https://github.com/eclipse/lemminx#clients and it gives the capability for a new consumer to know the community which uses your project).

In my side, we use the great kafkajs project inside [vscode-kafka](https://github.com/jlandersen/vscode-kafka) and it works like a charm! Thank a lot!

**Comment by mateusppereira** [comment]
using on production environments here in [Quero Education](querobolsa.com) 🇧🇷 🎉

**Comment by perniciachd** [comment]
I am very new to KafkaJS. I am able to write producers and consumers locally. But, for production deployment, I am not getting any idea of how to deploy either on Azure or AWS. If anyone has suggestions, please share.

**Comment by yarncraft** [comment]
> I am very new to KafkaJS. I am able to write producers and consumers locally. But, for production deployment, I am not getting any idea of how to deploy either on Azure or AWS. If anyone has suggestions, please share.

### Here is a list with possible deployment strategies:

Deploy with Cloud Managed Solution [Easy to maintain]:
- https://docs.microsoft.com/en-us/azure/event-hubs/event-hubs-for-kafka-ecosystem-overview
- https://www.confluent.io/confluent-cloud/
- https://aiven.io/kafka
- https://aws.amazon.com/msk/

Deploy with Docker [Single Zone]:
- https://www.baeldung.com/ops/kafka-docker-setup
- https://www.kaaiot.com/blog/kafka-docker

Deploy with Kubernetes [Scalable, Multi Zone]:
- https://github.com/banzaicloud/kafka-operator
- https://strimzi.io/
- https://bitnami.com/stack/kafka/helm

I hope this helps!

**Comment by AlexanderKaran** [comment]
Using it in Production at ClimateClever for communication between our microservices. KafkaJS plus Confluent Cloud integration took me half a day, love it

**Comment by LucasKeiti28** [comment]
we are using in e-commerce squad at [Dasa](https://dasa.com.br/)!

**Comment by HeeL** [comment]
@ [20 Minuten](https://www.20min.ch) we were able to achieve great performance and stable results for our analytics system that is now based on KafkaJS :v:

**Comment by feydan** [comment]
[ONE](https://www.onefinance.com/) (banking platform) has been using KafkaJs in our application since inception for asynchronous communication between bounded contexts, concurrent resilient processing of reactions to event log events, and to process external webhook and websocket event data (as well as make it durable).

**Comment by Durairaj** [comment]
A financial platform where it serves the trading applications, payment applications with over 100+ microservices on different projects. We use Kafkajs for over 2+ years for now.

**Comment by Saeger** [comment]
We @ [Digitalroute](https://www.digitalroute.com/) - a Swedish company based in Stockholm - has been using KafkaJS for almost a year now across multiple microservices for our Cloud SaaS product, and we're very happy with it :)

**Comment by harshagarwal00** [comment]
I agree with @ThisIsMissEm please add recipes it would be v useful, for example I found this recipe useful:
https://github.com/tulios/kafkajs/issues/808#issuecomment-922208802

**Comment by abhishandilya** [comment]
[GAS](https://greenactionstudio.com) is using it for EV charging software

**Comment by sandeeprapido** [comment]
[Rapido](https://www.rapido.bike/) is a mobility company in India.

**Comment by HeatherFlux** [comment]
We use it in prod with confluent cloud at [Extend](https://www.extend.com/)  it's used for our event first architecture with confluent. We've actually been building a lot of tooling with it and want to slowly open source some of the IAC tools we built to integrate serverless arch with confluent cloud.

**Comment by KoderPex** [comment]
We use KafkaJS + NodeJS with hexagonal architecture and we have a lot of problems with rebalances even had increased session timeout.
